app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt

```diff
@@ -0,0 +1,314 @@
+            }
+            wayIds.forEach { wayId ->
+                relationIdsByElementKeyCache[ElementKey(ElementType.WAY, wayId)]?.let { relationIds.addAll(it) }
+            }
+
+            wayIds.forEach {
+                val key = ElementKey(ElementType.WAY, it)
+                result.put(wayRelationCache[key]!!, wayRelationGeometryCache[key])
+            }
+            relationIds.forEach {
+                val key = ElementKey(ElementType.RELATION, it)
+                result.put(wayRelationCache[key]!!, wayRelationGeometryCache[key])
+            }
+        }
+
+        // todo: call trimNonSpatialCaches if tilesToFetch.isNotEmpty()?
+        //  we just added a lot of data to cache, so maybe we should drop something?
+        //  or maybe first check whether spatialCache is full?
+
+        return result
+    }
+
+    fun clear() {
+        synchronized(this) {
+            wayIdsByNodeIdCache.clear()
+            relationIdsByElementKeyCache.clear()
+            wayRelationCache.clear()
+            wayRelationGeometryCache.clear()
+            spatialCache.clear()
+        }
+    }
+
+    fun trim(size: Int) {
+        spatialCache.trim(size)
+        trimNonSpatialCaches()
+    }
+
+    private fun trimNonSpatialCaches() {
+        synchronized(this) {
+            val cachedNodeIds = spatialCache.getKeys().toHashSet()
+            // ways with at least one node in cache should not be removed
+            val waysWithCachedNode = hashSetOf<Long>()
+            // relations with at least one element in cache should not be removed
+            val relationsWithCachedElement = hashSetOf<Long>()
+            cachedNodeIds.forEach { nodeId ->
+                wayIdsByNodeIdCache[nodeId]?.let { waysWithCachedNode.addAll(it) }
+                relationIdsByElementKeyCache[ElementKey(ElementType.NODE, nodeId)]?.let { relationsWithCachedElement.addAll(it) }
+            }
+            waysWithCachedNode.forEach { wayId ->
+                relationIdsByElementKeyCache[ElementKey(ElementType.WAY, wayId)]?.let { relationsWithCachedElement.addAll(it) }
+            }
+            wayRelationCache.keys.retainAll {
+                if (it.type == ElementType.RELATION)
+                    it.id in relationsWithCachedElement
+                else it.id in waysWithCachedNode
+            }
+            wayRelationGeometryCache.keys.retainAll {
+                if (it.type == ElementType.RELATION)
+                    it.id in relationsWithCachedElement
+                else it.id in waysWithCachedNode
+            }
+
+            // now clean up wayIdsByNodeIdCache and relationIdsByElementKeyCache
+            wayIdsByNodeIdCache.keys.retainAll { it in cachedNodeIds }
+            relationIdsByElementKeyCache.keys.retainAll {
+                (it.type == ElementType.NODE && it.id in cachedNodeIds)
+                    || (it.type == ElementType.WAY && it.id in waysWithCachedNode)
+            }
```

Comment:
```suggestion
            val cachedNodeIds = spatialCache.getKeys().toHashSet()
            // ways with at least one node in cache should not be removed
            val wayIdsWithCachedNode = HashSet<Long>(cachedNodeIds.size / 2)
            // relations with at least one element in cache should not be removed
            val relationIdsWithCachedElement = HashSet<Long>(cachedNodeIds.size / 10)
            cachedNodeIds.forEach { nodeId ->
                wayIdsByNodeIdCache[nodeId]?.let { wayIdsWithCachedNode.addAll(it) }
                relationIdsByElementKeyCache[ElementKey(ElementType.NODE, nodeId)]?.let { relationIdsWithCachedElement.addAll(it) }
            }
            wayIdsWithCachedNode.forEach { wayId ->
                relationIdsByElementKeyCache[ElementKey(ElementType.WAY, wayId)]?.let { relationIdsWithCachedElement.addAll(it) }
            }
            wayRelationCache.keys.retainAll {
                if (it.type == ElementType.RELATION)
                    it.id in relationIdsWithCachedElement
                else it.id in wayIdsWithCachedNode
            }
            wayRelationGeometryCache.keys.retainAll {
                if (it.type == ElementType.RELATION)
                    it.id in relationIdsWithCachedElement
                else it.id in wayIdsWithCachedNode
            }

            // now clean up wayIdsByNodeIdCache and relationIdsByElementKeyCache
            wayIdsByNodeIdCache.keys.retainAll { it in cachedNodeIds }
            relationIdsByElementKeyCache.keys.retainAll {
                (it.type == ElementType.NODE && it.id in cachedNodeIds)
                    || (it.type == ElementType.WAY && it.id in wayIdsWithCachedNode)
            }
```
name vars so it is clear that it is about the ids, not the ways/relations themselves.
Also, following your own estimation, there is a small performance improvement to make by making the sets about the right size from the start,